sdk/core/src/main/java/org/apache/mesos/config/ConfigurationUpdater.java

```diff
@@ -0,0 +1,230 @@
+    private final Collection<ConfigurationValidator<C>> validators;
+
+    public ConfigurationUpdater(
+            StateStore stateStore,
+            ConfigStore<C> configStore,
+            ConfigurationComparer<C> configComparer,
+            Collection<ConfigurationValidator<C>> validators) {
+        this.stateStore = stateStore;
+        this.configStore = configStore;
+        this.configComparer = configComparer;
+        this.validators = validators;
+    }
+
+    /**
+     * Validates the provided {@code candidateConfig}, and updates the target configuration if the
+     * validation passes.
+     *
+     * @param candidateConfig the new candidate configuration to be validated and potentially marked
+     *                        as the target configuration
+     * @return a list of errors if the new candidate configuration failed validation, or an empty
+     *         list if the configuration passed validation and was updated to the config store
+     * @throws ConfigStoreException if there's an error when reading or writing to the config store,
+     *                              or if there are validation errors against the config and no
+     *                              prior fallback config is available
+     */
+    public UpdateResult updateConfiguration(C candidateConfig) throws ConfigStoreException {
+        // Get the currently stored target configuration
+        UUID targetConfigId;
+        try {
+            targetConfigId = configStore.getTargetConfig();
+        } catch (ConfigStoreException e) {
+            LOGGER.debug("No target configuration ID was set. First launch?");
+            targetConfigId = null;
+        }
+        C targetConfig;
+        if (targetConfigId != null) {
+            targetConfig = configStore.fetch(targetConfigId);
+        } else {
+            targetConfig = null;
+        }
+
+        // Log the config state before proceeding with checks.
+        final List<ConfigurationValidationError> errors = new ArrayList<>();
+        if (targetConfig != null) {
+            try {
+                LOGGER.info("Old target config: {}", targetConfig.toJsonString());
+            } catch (Exception e) {
+                LOGGER.error("Unable to get JSON representation of old target config object", e);
+                // Don't add a validation error: That'd prevent the new config from replacing this
+                // one, and we'd be stuck with this config forever! Hopefully the new config will fix things...
+            }
+        } else {
+            LOGGER.info("Old target config: <null>");
+        }
+        try {
+            LOGGER.info("New prospective config: {}", candidateConfig.toJsonString());
+        } catch (Exception e) {
+            LOGGER.error("Unable to get JSON representation of new prospective config object: " + candidateConfig, e);
+            errors.add(ConfigurationValidationError.valueError("NewConfigAsJson", "jsonString",
```

Comment:
Should we short-circuit here?  We continue to attempt validation anyway right now.